tests/overlay_tests.cpp

```diff
@@ -1981,7 +1981,129 @@ TEST_F(OverlayTest, supervisor)
+
+  // Sanity checks of overlay networking configuration.
+  ASSERT_SOME(state);
+  ASSERT_EQ(1, state->network().overlays_size());
+  ASSERT_EQ(OVERLAY_NAME, state->network().overlays(0).name());
+  ASSERT_EQ(OVERLAY_SUBNET, state->network().overlays(0).subnet());
+  ASSERT_EQ(OVERLAY_SUBNET6, state->network().overlays(0).subnet6());
+  ASSERT_EQ(OVERLAY_PREFIX, state->network().overlays(0).prefix());
+  ASSERT_EQ(OVERLAY_PREFIX6, state->network().overlays(0).prefix6());
+
+  // Make sure that no agents are registered yet.
+  ASSERT_EQ(0, state->agents_size());
+
+  AgentConfig agentOverlayConfig;
+  agentOverlayConfig.set_master(stringify(overlayMaster.address));
+
+  // Setup a future to notify the test that Agent overlay module has
+  // registered.
+  Future<RegisterAgentMessage> registerAgentMessage =
+    FUTURE_PROTOBUF(RegisterAgentMessage(), _, _);
+
+  // Setup a future to notify the test that Agent overlay module has
+  // registered.
+  Future<AgentRegisteredMessage> agentRegisteredMessage =
+    FUTURE_PROTOBUF(AgentRegisteredMessage(), _, _);
+
+  Try<Owned<overlayAgent::ManagerProcess>> agentModule = startOverlayAgent(
+      agentOverlayConfig);
+
+  ASSERT_SOME(agentModule);
+  AWAIT_READY(registerAgentMessage);
+  AWAIT_READY(agentRegisteredMessage);
+
+  // Check that the agent is allowed to progress.
+  AWAIT_READY(agentModule.get()->ready());
+
+  // Hit the `state` endpoint again to verify that the agent has been
+  // registered.
+  stateResponse = process::http::get(
+      overlayMaster,
+      "state");
+
+  AWAIT_EXPECT_RESPONSE_STATUS_EQ(OK().status, stateResponse);
+  AWAIT_EXPECT_RESPONSE_HEADER_EQ(
+      APPLICATION_JSON,
+      "Content-Type",
+      stateResponse);
+
+  state = parseMasterState(stateResponse->body);
+  ASSERT_SOME(state);
+  ASSERT_EQ(1, state->agents_size());
+
+  // Now let's drop the agent record from the state.
+  string requestBody = "{\"agents\":[\"" + state->agents(0).ip() + "\"]}";
+  stateResponse = process::http::post(
+       overlayMaster,
+       "agents/delete",
+       None(),
+       requestBody,
+       "application/json");
+
+  AWAIT_EXPECT_RESPONSE_STATUS_EQ(OK().status, stateResponse);
+  AWAIT_EXPECT_RESPONSE_HEADER_EQ(
+      APPLICATION_JSON,
+      "Content-Type",
+      stateResponse);
+
+  // We get the full state in response. Let's verify that there are
+  // no agents registered.
+  state = parseMasterState(stateResponse->body);
+  ASSERT_SOME(state);
+  ASSERT_EQ(0, state->agents_size());
+
+
+  // Double-check that the agent is gone from the state by hitting
+  // the `state` endpoint once again.
+  stateResponse = process::http::get(
+      overlayMaster,
+      "state");
+
+  AWAIT_EXPECT_RESPONSE_STATUS_EQ(OK().status, stateResponse);
+  AWAIT_EXPECT_RESPONSE_HEADER_EQ(
+      APPLICATION_JSON,
+      "Content-Type",
+      stateResponse);
+
+  state = parseMasterState(stateResponse->body);
+  ASSERT_SOME(state);
+  ASSERT_EQ(0, state->agents_size());
```

Comment:
Could you please also add a test case that checks if it's possible to allocate IP that was previously deallocated?

Reply:
Done. Please check it out.

Reply:
I didn't find that it checks if the same subnet is allocated. Did I overlook it?

Reply:
You mean, that it gets exactly the same subnet that it got the first time it was tilted up? There is no such guarantee, and AFAIU, there can't be any since it depends on the order in which agent registration requests come in.

Reply:
For testing purpose we can have a very small subnet for all agents that allows to have only one agent node to be allocated. I'd like to have some test coverage to the whole use case, it's not just deallocation, but also re-use of deallocated subnets by agent nodes with new node ip addresses.